client-apis/decrypt.md

```diff
@@ -69,65 +98,259 @@ This default CMM MUST obtain the decryption materials required for decryption.
+  this operation MAY immediately complete and succeed,
+  or it MAY wait until an end of the input bytes is indicated
+  to complete and succeed.
+- If this operation successfully completes the above steps
+  but there are available bytes which have not been processed,
+  this operation MUST fail.
+
+### Parse the header
+
+Given encrypted message bytes, this operation MUST process those bytes sequentially,
+deserializing those bytes according to the [message format](../data-format/message.md).
+
+This operation MUST attempt to deserialize all available encrypted message bytes until it has
+successfully deserialized a valid [message header](../data-format/message-header.md).
+
+This operation MUST wait if it doesn't have enough encrypted message bytes available to
+deserialize the next field of the message header until enough input bytes become available or
+the caller indicates an end to the encrypted message.
+
+Until the [header is verified](#verify-the-header), this operation MUST NOT
+release any parsed information from the header.
+
+### Get the decryption materials
+
+To verify the message header and decrypt the message body,
+a set of valid decryption materials is required.
+
+This operation MUST obtain this set of [decryption materials](../framework/structures.md#decryption-materials),
+by calling [Decrypt Materials](../framework/cmm-interface.md#decrypt-materials) on a [CMM](../framework/cmm-interface.md).
+
+The CMM used MUST be the input CMM, if supplied.
+If a CMM is not supplied as the input, the decrypt operation MUST construct a [default CMM](../framework/default-cmm.md)
+from the [keyring](../framework/keyring-interface.md) inputted.
+
+The call to the CMM's [Decrypt Materials](../framework/cmm-interface.md#decrypt-materials) operation
+MUST be constructed as follows:
+
+- Encryption Context: This is the parsed [encryption context](../data-format/message-header.md#aad)
+  from the message header.
+- Algorithm Suite ID: This is the parsed
+  [algorithm suite ID](../data-format/message-header.md#algorithm-suite-id)
+  from the message header.
+- Encrypted Data Keys: This is the parsed [encrypted data keys](../data-format/message-header#encrypted-data-keys)
+  from the message hedaer.
+
+The data key used as input for all decryption described below is a data key derived from the plaintext data key
+included in the [decryption materials](../framework/structures.md#decryption-materials).
+The algorithm used to derive a data key from the plaintext data key MUST be
+the [key derivation algorithm](../framework/algorithm-suites.md#key-derivation-algorithm) included in the
+[algorithm suite](../framework/algorithm-suites.md) defined above.
+This document refers to the output of the key derivation algorithm as the derived data key.
+Note that if the key derivation algorithm is the [identity KDF](../framework/algorithm-suites.md#identity-kdf),
+then the derived data key is the same as the plaintext data key.
+
+### Verify the header
```

Comment:
This and the Body verification should be moved to the Algorithm Suites, but that can be done in a later feature request.

Reply:
Big +1 from me for factoring as much as possible into Algorithm Suites to shorten these features. If possible to do in this PR I'd find it easier to review and sign-off on the whole package.

Reply:
As agreed offline, we will do this in a separate PR scoped to spec 1.1. Created issue here: https://github.com/awslabs/aws-encryption-sdk-specification/issues/162